### PR TITLE
[Navi21][COMGR][HIPRTC][WORKAROUND] W/A for `warpSize` issue when PCH is used.

### PR DESCRIPTION
**Pull request by atamazov**
This is W/A for #1431 which replaces #1436. :yellow_circle: It allows to reduce urgency of #1431 down to `normal`.

- Reverted #1436.
  - :information_source: I recommend skipping the very first commit during reviewing.
    - [Click here to see how it looks](https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1469/files/f4ad5cf205ffab6566db34ffd1dca79bd8f51f4a..562eaf89b748f6104888557ebc81cd7bd708d10b).
- In all HIP kernels that use `warpSize`, the W/A is applied as described below.
- :red_circle: Also fixes potential issue (applicable only to HIP builds via COMGR):
  - _Device libs for wavesize 64 are used for gfx10 targets even if wavesize is 32._

## :warning: How `warpSize` must be used in HIP sources

From now on (until the `WORKAROUND_ISSUE_1431` macro is removed from MIOpen host code), any HIP kernel that uses predefined symbol `warpSize` must include the following code snippet:
```C++
#ifdef WORKAROUND_ISSUE_1431
#include "workaround_issue_1431.hpp"
#endif
```
No other changes needed.

### CODE REVIEW

**Comment by atamazov** [comment]
@junliume Please assign reviewers and other attributes (can't do that, sorry).

**Comment by atamazov** [comment]
@junliume If we are going to disable HIPRTC very soon, then we need to remove WORKAROUND_ISSUE_898 from tests/CMakeLists.txt.

Why: When PCH is disabled, the newly added regression test does nothing useful. WORKAROUND_ISSUE_898 disables PCH for HIP builds via COMGR. Right now this is not a problem, because we build HIP kernels via HIPRTC (and WORKAROUND_ISSUE_898 does not affect HIPRTC builds). But the regression test will be useless as soon as #1470 is merged in.

Please let me know if you would like me to remove W/A for issue 898 right in this PR or separately. See https://github.com/ROCmSoftwarePlatform/MIOpen/issues/898#issuecomment-1036629696 for more info.

**Comment by junliume** [comment]
> @junliume If we are going to disable HIPRTC very soon, then we need to remove WORKAROUND_ISSUE_898 from tests/CMakeLists.txt.
> 
> Why: When PCH is disabled, the newly added regression test does nothing useful. WORKAROUND_ISSUE_898 disables PCH for HIP builds via COMGR. Right now this is not a problem, because we build HIP kernels via HIPRTC (and WORKAROUND_ISSUE_898 does not affect HIPRTC builds). But the regression test will be useless as soon as #1470 is merged in.
> 
> Please let me know if you would like me to remove W/A for issue 898 right in this PR or separately. See [#898 (comment)](https://github.com/ROCmSoftwarePlatform/MIOpen/issues/898#issuecomment-1036629696) for more info.

Could you help with the QA removal or let @DrizztDoUrden work on it? Thanks!

**Review by qianfengz — DISMISSED** [review]

**Comment by atamazov** [comment]
@qianfengz 
> I have tested the branch `atamazov:workaround-warpsize-2` and verified that `bin/test_reduce_test --all` can pass on ROCM 4.3.1/gfx1030

Thanks for double checking this ;)

**Comment by atamazov** [comment]
@DrizztDoUrden @junliume Can you please let me know CI status (and push this through CI if necessary), Thanks!

**Comment by DrizztDoUrden** [comment]
I have started it.

**Review by shaojiewang — COMMENTED** [review]
Code looks good.
But CI is blocked by hip-tidy.
```
[ 70%] clang-tidy: Running clang-tidy on target solver/batchnorm/forward_inference.cpp...

[2022-03-25T18:23:23.375Z] /var/jenkins/workspace/MLLibs_MIOpen_PR-1469/src/comgr.cpp:841:28: error: use emplace_back instead of push_back [hicpp-use-emplace,modernize-use-emplace,-warnings-as-errors]

[2022-03-25T18:23:23.375Z]                 optCompile.push_back("-DWORKAROUND_ISSUE_1431=1");

[2022-03-25T18:23:23.375Z]                            ^~~~~~~~~~

[2022-03-25T18:23:23.375Z]                            emplace_back(

[2022-03-25T18:23:23.375Z] src/CMakeFiles/tidy-target-MIOpen-comgr_cpp.dir/build.make:57: recipe for target 'src/CMakeFiles/tidy-target-MIOpen-comgr_cpp' failed
```

**Comment by junliume** [comment]
> Code looks good. But CI is blocked by hip-tidy.
> 
> ```
> [ 70%] clang-tidy: Running clang-tidy on target solver/batchnorm/forward_inference.cpp...
> 
> [2022-03-25T18:23:23.375Z] /var/jenkins/workspace/MLLibs_MIOpen_PR-1469/src/comgr.cpp:841:28: error: use emplace_back instead of push_back [hicpp-use-emplace,modernize-use-emplace,-warnings-as-errors]
> 
> [2022-03-25T18:23:23.375Z]                 optCompile.push_back("-DWORKAROUND_ISSUE_1431=1");
> 
> [2022-03-25T18:23:23.375Z]                            ^~~~~~~~~~
> 
> [2022-03-25T18:23:23.375Z]                            emplace_back(
> 
> [2022-03-25T18:23:23.375Z] src/CMakeFiles/tidy-target-MIOpen-comgr_cpp.dir/build.make:57: recipe for target 'src/CMakeFiles/tidy-target-MIOpen-comgr_cpp' failed
> ```

@atamazov  I cannot change the branch directly so could you suppress this warning?
[        -hicpp-use-emplace -modernize-use-emplace](https://github.com/ROCmSoftwarePlatform/MIOpen/blob/6f08ed32086b3a0f7a311650bc994ac0ef093932/CMakeLists.txt#L550)

I have to do it in another PR https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1504 too. Unless the perf gain is visible in a benchmark, I do not really think we should update/revise all of them.

**Comment by junliume** [comment]
> > Code looks good. But CI is blocked by hip-tidy.
> > ```
> > [ 70%] clang-tidy: Running clang-tidy on target solver/batchnorm/forward_inference.cpp...
> > 
> > [2022-03-25T18:23:23.375Z] /var/jenkins/workspace/MLLibs_MIOpen_PR-1469/src/comgr.cpp:841:28: error: use emplace_back instead of push_back [hicpp-use-emplace,modernize-use-emplace,-warnings-as-errors]
> > 
> > [2022-03-25T18:23:23.375Z]                 optCompile.push_back("-DWORKAROUND_ISSUE_1431=1");
> > 
> > [2022-03-25T18:23:23.375Z]                            ^~~~~~~~~~
> > 
> > [2022-03-25T18:23:23.375Z]                            emplace_back(
> > 
> > [2022-03-25T18:23:23.375Z] src/CMakeFiles/tidy-target-MIOpen-comgr_cpp.dir/build.make:57: recipe for target 'src/CMakeFiles/tidy-target-MIOpen-comgr_cpp' failed
> > ```
> 
> @atamazov I cannot change the branch directly so could you suppress this warning? [ -hicpp-use-emplace -modernize-use-emplace](https://github.com/ROCmSoftwarePlatform/MIOpen/blob/6f08ed32086b3a0f7a311650bc994ac0ef093932/CMakeLists.txt#L550)
> 
> I have to do it in another PR #1504 too. Unless the perf gain is visible in a benchmark, I do not really think we should update/revise all of them.

Ping @atamazov

File: src/comgr.cpp
Comment (junliume) [inline src/comgr.cpp:841]:
```suggestion
                optCompile.emplace_back("-DWORKAROUND_ISSUE_1431=1");
```

File: src/comgr.cpp
Comment (junliume) [inline src/comgr.cpp:852]:
```suggestion
                addDevLibs.emplace_back("wavefrontsize64");
```

**Comment by atamazov** [comment]
Dear reviewers, can you please check CI status? Thanks!

File: src/kernels/workaround_issue_1431.hpp
Comment (junliume) [inline src/kernels/workaround_issue_1431.hpp:35]: @atamazov @JehandadKhan  it seems an order issue, still conflicting with `hip_runtime.h`

```
[2022-04-12T02:44:44.578Z] ../bin/test_reduce_test --half --D 7 1024 1 --R 0 1 2 --ReduceOp 0 --CompType 1 --N 0 --I 0 --scales 1 0 
[2022-04-12T02:44:44.578Z] FAILED: /var/jenkins/workspace/MLLibs_MIOpen_PR-1469/src/hipoc/hipoc_program.cpp:300: Code object build failed. Source: gridwise_generic_reduction_second_call_warpwise_reduce_all_dims.cpp
[2022-04-12T02:44:44.578Z] verify_reduce_no_indices failed
[2022-04-12T02:44:44.578Z] Input Tensor 7, 1024, 1
[2022-04-12T02:44:44.578Z] MIOpen(HIP): Error [Do] 'amd_comgr_do_action(kind, handle, in.GetHandle(), out.GetHandle())' AMD_COMGR_ACTION_COMPILE_SOURCE_TO_BC: ERROR (1)
[2022-04-12T02:44:44.578Z] MIOpen(HIP): Error [BuildHip] comgr status = ERROR (1)
[2022-04-12T02:44:44.578Z] MIOpen(HIP): Warning [BuildHip] In file included from /tmp/comgr-843102/input/gridwise_generic_reduction_second_call_warpwise_reduce_all_dims.cpp:29:
[2022-04-12T02:44:44.578Z] In file included from /tmp/comgr-843102/include/config.hpp:5:
[2022-04-12T02:44:44.578Z] In file included from /opt/rocm-4.3.1/hip/../include/hip/hip_runtime.h:62:
[2022-04-12T02:44:44.578Z] In file included from /opt/rocm-4.3.1/hip/../include/hip/amd_detail/hip_runtime.h:63:
[2022-04-12T02:44:44.578Z] /opt/rocm-4.3.1/hip/../include/hip/hip_runtime_api.h:89:9: error: expected member name or ';' after declaration specifiers
[2022-04-12T02:44:44.578Z]     int warpSize;              ///< Warp size.
[2022-04-12T02:44:44.578Z]     ~~~ ^
[2022-04-12T02:44:44.578Z] /tmp/comgr-843102/include/workaround_issue_1431.hpp:35:18: note: expanded from macro 'warpSize'
[2022-04-12T02:44:44.578Z] #define warpSize 32
[2022-04-12T02:44:44.578Z]                  ^
[2022-04-12T02:44:44.578Z] In file included from /tmp/comgr-843102/input/gridwise_generic_reduction_second_call_warpwise_reduce_all_dims.cpp:29:
[2022-04-12T02:44:44.578Z] In file included from /tmp/comgr-843102/include/config.hpp:5:
[2022-04-12T02:44:44.578Z] In file included from /opt/rocm-4.3.1/hip/../include/hip/hip_runtime.h:62:
[2022-04-12T02:44:44.578Z] In file included from /opt/rocm-4.3.1/hip/../include/hip/amd_detail/hip_runtime.h:70:
[2022-04-12T02:44:44.578Z] In file included from /opt/rocm-4.3.1/hip/../include/hip/amd_detail/hip_atomic.h:3:
[2022-04-12T02:44:44.578Z] /opt/rocm-4.3.1/hip/../include/hip/amd_detail/device_functions.h:295:22: error: expected unqualified-id
[2022-04-12T02:44:44.578Z] static constexpr int warpSize = __AMDGCN_WAVEFRONT_SIZE;
[2022-04-12T02:44:44.578Z]                      ^
[2022-04-12T02:44:44.578Z] /tmp/comgr-843102/include/workaround_issue_1431.hpp:35:18: note: expanded from macro 'warpSize'
[2022-04-12T02:44:44.578Z] #define warpSize 32
[2022-04-12T02:44:44.578Z]                  ^
[2022-04-12T02:44:44.578Z] 2 errors generated when compiling for gfx1030.

Comment (junliume) [reply src/kernels/workaround_issue_1431.hpp:35]: @atamazov It seem to be the same issues discussed at https://github.com/ROCmSoftwarePlatform/MIOpen/issues/1431

**Comment by junliume** [comment]
**Update all ROCm 5.2 targeted PR/issues as blockers.**
this workaround has issues with comgr and may not land within time frame.

File: src/kernels/workaround_issue_1431.hpp
Comment (atamazov) [reply src/kernels/workaround_issue_1431.hpp:35]: The W/A in this PR is working incorrectly when `MIOPEN_DEBUG_COMGR_HIP_PCH_ENFORCE=0`.

The reason is a coding bug in my PR. W/A for issue #1431 should not be enabled when PCH is not enabled. Currently W/A is not enabled when _"PCH is not supported"_ but this is not the same as _"PCH is not enabled"_. For example, PCH is supported but not enabled when `MIOPEN_DEBUG_COMGR_HIP_PCH_ENFORCE=0` in the environment.

It seems like the failing test is running with `MIOPEN_DEBUG_COMGR_HIP_PCH_ENFORCE=0` (it comes from https://github.com/atamazov/MIOpen/blob/c330fd9ec8c4cdf77783ffd5f96d08bf37f34d9b/test/CMakeLists.txt#L277 from my fork, which is outdated and still contains W/A for issue #898 from where this env var came from).

I'll fix the bug and update PR from the recent develop.

Comment (junliume) [reply src/kernels/workaround_issue_1431.hpp:35]: #1508 is already merged so maybe update your branch with the latest develop?

Comment (atamazov) [reply src/kernels/workaround_issue_1431.hpp:35]: Fix is also required. Please expect updated PR tomorrow.

Comment (junliume) [reply src/kernels/workaround_issue_1431.hpp:35]: @atamazov  last normal promotion should be started today so I think this one would likely to be a cherry picking candidate instead. :)

Comment (atamazov) [reply src/kernels/workaround_issue_1431.hpp:35]: Understood, thanks. BTW fix is in developement/testing still...

Comment (atamazov) [reply src/kernels/workaround_issue_1431.hpp:35]: Resolved in 65ca9f2229dd6885cc83eb316f25cc2d88fa70a7

**Comment by atamazov** [comment]
@junliume @shaojiewang @qianfengz Ready for review.

**Review by junliume — APPROVED** [review]

**Pull request merged**
